drivers/mfd/tssupervisor-core.c

```diff
@@ -119,6 +121,66 @@ static ssize_t wake_en_show(struct device *dev,
 }
 static DEVICE_ATTR_RW(wake_en);
 
+static ssize_t serial_store(struct device *dev, struct device_attribute *attr,
+			      const char *buf, size_t count)
+{
+	struct ts_supervisor *super = dev_get_drvdata(dev);
+	unsigned int ctrl = 0;
+	u8 sn[6];
+	int ret;
+
+	ret = regmap_read(super->regmap, SUPER_SERIAL_CTRL, &ctrl);
+	if (ret)
+		return ret;
+
+	/* Dont write the SN if its already written */
+	if (ctrl & SUPER_SN_LOCKED)
+		return -EEXIST;
+
+	if (count < 17)
+		return -EIO;
+
+	if (sscanf(buf, "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+		   &sn[0], &sn[1], &sn[2], &sn[3], &sn[4], &sn[5]) != 6)
+		return -EINVAL;
+
+	ret = regmap_bulk_write(super->regmap, SUPER_SERIAL0, sn, 3);
+	if (ret)
+		return ret;
+
+	ret = regmap_update_bits(super->regmap, SUPER_SERIAL_CTRL,
```

Comment:
Maybe add a readback check to ensure the data we expect to be written was written before locking out future writes?

Reply:
The actual locking out of future writes happens on the microcontroller if the write succeeds.  If it were to fail to write for some reason then the SN_LOCKED bit would not be set the next time its read.  If you mean to double check the i2c write, we should fail out before them if the bulk_write fails before locking out the mac address.

Reply:
So, if there is a bit flip on I2C, we have no way to read back what we tried to write and check that the written SN is what was intended to be written and correct if there was an issue?

Reply:
Just force pushed a change for this to check the OTP data before writing it.